File: Framework_Poison-Rat/rat.ts
import * as crypto from 'crypto';
import * as net from 'net';

const HOST = '127.0.0.1';
const PORT = 443;

const chave = crypto.createHash('sha256').update('your-secret-key').digest();
const ivSize = 12; // Tamanho do IV para AES-GCM

// Função para gerar um IV aleatório
function gerarIV(): Buffer {
    return crypto.randomBytes(ivSize);
}

// Função para enviar pacotes criptografados
function enviarPacoteCriptografado(socket: net.Socket, mensagem: string) {
    const tamanhoPacote = 512;
    let offset = 0;

    while (offset < mensagem.length) {
        const remaining = mensagem.length - offset;
        const chunkSize = Math.min(tamanhoPacote, remaining);
        const chunk = mensagem.slice(offset, offset + chunkSize);

        // Adiciona um contador ao pacote para evitar repetição
        const pacoteComNonce = `${offset}|${chunk}`;

        // Criptografa com AES-256-GCM
        const iv = gerarIV();
        const cipher = crypto.createCipheriv('aes-256-gcm', chave, iv);
        let pacoteCriptografado = cipher.update(pacoteComNonce, 'utf8', 'base64');
        pacoteCriptografado += cipher.final('base64');

        // Anexa IV e Tag de autenticação
        const pacoteFinal = iv.toString('base64') + cipher.getAuthTag().toString('base64') + pacoteCriptografado;
        socket.write(pacoteFinal);

        offset += chunkSize;
    }
}

// Criar conexão com o servidor
const client = new net.Socket();
client.connect(PORT, HOST, () => {
    console.log('Conectado ao servidor');

    // Exemplo de mensagem a ser enviada
    const mensagem = ".";
    enviarPacoteCriptografado(client, mensagem);

    // Manter a comunicação aberta e ler resposta do servidor enquanto estiver ativo
    client.on('data', (data) => {
        console.log(data.toString());
    });

    client.on('end', () => {
        console.log('Desconectado do servidor');
    });
});

File: Framework_Poison-Rat/rat.swift
import CryptoKit
import Foundation
import Network

let host = "127.0.0.1"
let port: UInt16 = 443
let chave = "your-secret-key-32-bytes".data(using: .utf8)!

// Função para gerar um IV aleatório
func gerarIV() -> Data {
    var iv = Data(count: 12)
    let _ = iv.withUnsafeMutableBytes { (bytes: UnsafeMutableRawBufferPointer) in
        let result = SecRandomCopyBytes(kSecRandomDefault, iv.count, bytes.baseAddress!)
        precondition(result == errSecSuccess, "falha ao gerar IV")
    }
    return iv
}

// Função para enviar pacotes criptografados
func enviarPacoteCriptografado(to connection: NWConnection, mensagem: String) {
    let tamanhoPacote = 512
    var offset = 0

    while offset < mensagem.count {
        let remaining = mensagem.count - offset
        let chunkSize = min(tamanhoPacote, remaining)
        let chunk = String(mensagem[mensagem.index(mensagem.startIndex, offsetBy: offset)..<mensagem.index(mensagem.startIndex, offsetBy: offset + chunkSize)])

        // Adiciona um contador ao pacote para evitar repetição
        let pacoteComNonce = "\(offset)|\(chunk)"

        // Criptografa com AES-256-GCM
        let iv = gerarIV()
        let aes = try! AES.GCM.seal(pacoteComNonce.data(using: .utf8)!, using: SymmetricKey(data: chave), nonce: AES.GCM.Nonce(data: iv))

        // Anexa IV e Tag de autenticação
        let pacoteFinal = iv + aes.ciphertext
        connection.send(content: pacoteFinal, completion: .contentProcessed { sendError in
            if let error = sendError {
                print("Erro ao enviar pacote:", error)
            }
        })

        offset += chunkSize
    }
}

// Criar uma conexão com o servidor
let connection = NWConnection(host: NWEndpoint.Host(host), port: NWEndpoint.Port(integerLiteral: port), using: .tls)

connection.start(queue: .main)
connection.stateUpdateHandler = { newState in
    switch newState {
    case .ready:
        print("Conectado ao servidor")
        
        // Exemplo de mensagem a ser enviada
        let mensagem = "."
        enviarPacoteCriptografado(to: connection, mensagem: mensagem)

        // Manter a comunicação aberta e ler resposta do servidor enquanto estiver ativo
        connection.receiveMessage { (data, context, isComplete, error) in
            if let error = error {
                print("Erro ao receber resposta:", error)
                return
            }

            if let data = data, !data.isEmpty {
                print("Resposta do servidor:", String(data: data, encoding: .utf8) ?? "")
            }

            // Continue recebendo enquanto a conexão estiver aberta
            connection.receiveMessage { (data, context, isComplete, error) in
                if isComplete {
                    print("Conexão encerrada.")
                    connection.cancel()
                }
            }
        }

    case .failed(let error):
        print("Falha na conexão:", error)
        connection.cancel()
    default:
        break
    }
}
